admin-dev/themes/new-theme/js/components/pagination/dynamic-grid-paginator.js

```diff
@@ -0,0 +1,246 @@
+      if (e.which === 13) {
+        e.preventDefault();
+        const page = this.validatePageNumber(Number(e.currentTarget.value));
+        const limit = this.getLimit();
+        this.gridRenderer.render(page, limit);
+        this.updatePaginator(page, limit);
+      }
+    });
+  }
+
+  /**
+   * @param {Number} currentPage
+   * @param {Number} limit
+   */
+  updatePaginator(currentPage, limit) {
+    this.setCurrentPageInput(currentPage);
+    if (currentPage === this.getPagesCount()) {
+      this.updatePaginatorForLastPage();
+    } else if (currentPage === 1) {
+      this.updatePaginatorForFirstPage();
+    } else {
+      this.updatePaginatorForMiddlePage(currentPage);
+    }
+  }
+
+  getLimit() {
+    return this.$paginationContainer.data('limit');
+  }
+
+  getPagesCount() {
+    const limit = this.getLimit();
+    const total = this.getTotal();
+
+    return Math.ceil(total / limit);
+  }
+
+  getTotal() {
+    return this.$paginationContainer.data('total');
+  }
+
+  validatePageNumber(page) {
+    if (page > this.getPagesCount()) {
+      return this.getPagesCount();
+    }
+
+    if (page < 1) {
+      return 1;
+    }
+
+    return page;
+  }
+
+  extractPageFromDataset(target) {
+    return Number($(target).data('page'));
+  }
+
+  setCurrentPageInput(page) {
+    $(this.selectorsMap.jumpToPageInput).val(page);
+  }
+
+  updatePaginatorForFirstPage() {
+    this.toggleFirstPageAvailability(false);
+    this.togglePreviousPageAvailability(false);
+    this.updateNextPageBtnData(2);
+    this.toggleNextPageAvailability(true);
+    this.toggleLastPageAvailability(true);
+  }
+
+  updatePaginatorForLastPage() {
+    this.toggleFirstPageAvailability(true);
+    this.toggleNextPageAvailability(false);
+    this.toggleLastPageAvailability(false);
+    this.updatePreviousPageBtnData(this.getPagesCount() - 1);
+    this.togglePreviousPageAvailability(true);
+  }
+
+  updatePaginatorForMiddlePage(page) {
+    this.updatePreviousPageBtnData(page - 1);
+    this.updateNextPageBtnData(page + 1);
+    this.toggleFirstPageAvailability(true);
+    this.togglePreviousPageAvailability(true);
+    this.toggleNextPageAvailability(true);
+    this.toggleLastPageAvailability(true);
+  }
+
+  updatePreviousPageBtnData(page) {
+    const previousPageBtn = this.$paginationContainer.find(this.selectorsMap.previousPageBtn);
+    previousPageBtn.data('url', this.router.generate(this.routing.route, this.mergeParamsToKeep({
+      [this.routing.pageKey]: page,
+      [this.routing.limitKey]: this.getLimit(),
+    })));
+    previousPageBtn.data('page', page);
+  }
+
+  updateNextPageBtnData(page) {
+    const nextPageBtn = this.$paginationContainer.find(this.selectorsMap.nextPageBtn);
+    nextPageBtn.data('url', this.router.generate(this.routing.route, this.mergeParamsToKeep({
+      [this.routing.pageKey]: page,
+      [this.routing.limitKey]: this.getLimit(),
+    })));
+    nextPageBtn.data('page', page);
+  }
```

Comment:
Same here, you could do it with only one method, with the selector as a param

Reply:
extracted duplicate code to separate method, but still left verbose calls, I think they help to understand what is happening even though the class gets bigger

Reply:
Don't you think adding comments helps enough ?